### PR TITLE
[AIR] Refactor `_get_unique_value_indices`

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Refactors `_get_unique_value_indices` (used in Encoder preprocessors) for much improved performance with multiple columns. Also uses the same, more robust intermediary dataset format in `_get_most_frequent_values` (Imputers).

The existing unit tests pass, and no functionality has been changed.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by clarkzinzow — APPROVED** [review]
LGTM, just a small type annotation nit!

File: python/ray/ml/preprocessors/imputer.py [hunk 1]
```diff
@@ -82,16 +82,17 @@ def _get_most_frequent_values(
 ) -> Dict[str, Union[str, Number]]:
     columns = list(columns)
 
-    def get_pd_value_counts(df: pd.DataFrame) -> List[Counter]:
-        return [Counter(df[col].value_counts().to_dict()) for col in columns]
+    def get_pd_value_counts(df: pd.DataFrame) -> List[Dict[str, Counter]]:
+        return [{col: Counter(df[col].value_counts().to_dict()) for col in columns}]
```
Comment (matthewdeng) [inline python/ray/ml/preprocessors/imputer.py:86]: Does this still need to be wrapped in an array? If so do we need to unwrap it in line 92?

File: python/ray/ml/preprocessors/encoder.py [hunk 2]
```diff
@@ -118,37 +118,40 @@ def __repr__(self):
 
 
 def _get_unique_value_indices(
-    dataset: Dataset, *columns: str
+    dataset: Dataset,
+    *columns: str,
+    drop_na_values: bool = False,
```
Comment (matthewdeng) [inline python/ray/ml/preprocessors/encoder.py:123]: Do we need to add this if it's not being used?

Comment (Yard1) [reply python/ray/ml/preprocessors/encoder.py:123]: It's actually being used in the Categorizer PR! Left it in here by mistake, but if we are going to merge that as well, it doesn't really make a huge difference to keep it in.

**Pull request merged**